### ISSUE TITLE
Subflow context (node/flow) is not reflected in «Context Data» tab

### ISSUE THREAD

**Issue by mu3** [node-red/node-red#2589]
### What are the steps to reproduce?

```
[{"id":"a2576be0.69ed68","type":"subflow","name":"Subflow 1","info":"","in":[{"x":50,"y":30,"wires":[{"id":"20e27550.2867aa"},{"id":"a66758e9.083618"}]}],"out":[]},{"id":"20e27550.2867aa","type":"change","z":"a2576be0.69ed68","d":true,"name":"","rules":[{"t":"set","p":"payload","pt":"flow","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":300,"y":160,"wires":[[]]},{"id":"a66758e9.083618","type":"function","z":"a2576be0.69ed68","name":"","func":"context.set('foo', 'bar')         // not showing up\nflow.set('foo', 'bar')            // not showing up\nflow.set('$parent.foo', 'bar')    // ok\nglobal.set('foo', 'bar')          // ok\n","outputs":1,"noerr":0,"x":270,"y":220,"wires":[[]]},{"id":"9125728e.3a84e","type":"subflow:a2576be0.69ed68","z":"e4f46c53.4616a","name":"","env":[],"x":180,"y":300,"wires":[]},{"id":"d07025c6.33da38","type":"inject","z":"e4f46c53.4616a","name":"","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":180,"y":260,"wires":[["9125728e.3a84e"]]}]
```

### What happens?

    context.set('foo', 'bar')         // not showing up
    flow.set('foo', 'bar')            // not showing up
    flow.set('$parent.foo', 'bar')    // ok
    global.set('foo', 'bar')          // ok

<img width="312" alt="Screen Shot 2020-06-02 at 4 05 40" src="https://user-images.githubusercontent.com/661220/83468707-55c88900-a486-11ea-83d6-628c8fb3a524.png">

### Please tell us about your environment:

- [ ] Node-RED version: 1.0.6
- [ ] Node.js version: 12.11.1; 10
- [ ] Platform/OS: macOS, Linux

**Comment by knolleary** [comment]
Yup - this is a long standing issue that the context sidebar doesn't play well with subflows.

But the issue is a bit more complex.

When you are viewing a subflow in the editor, you are viewing the template of the subflow, not a specific instance. There could be 10+ instances of the subflow in your workspace. When you select a node inside a subflow, there would then be 10 + separate copies of that node running - each with its own context data. That is why the editor can't display the context data for a node inside the subflow - it doesn't know which instance of the subflow you want to look at.

**Comment by mu3** [comment]
Yeah that makes sense 🙂
Thanks for clarification!

**Comment by fvbommel** [comment]
> That is why the editor can't display the context data for a node inside the subflow - it doesn't know which instance of the subflow you want to look at.

Could it not simply show up in the "Node" section when the subflow instance is selected?
Subflow context is specific to the subflow instance which is represented by that node, so it seems to fit. It might not be very handy while editing the subflow, but it'd be better than not having it available at all, which is the current situation...

Unless I'm missing something and there's some other node-specific context that might be more relevant for subflow nodes?

Renaming the "Node" section to "Subflow" in that case would also be a nice touch, but I'd say it's optional (especially if the documentation mentions where to find that context)

**Comment by prof7bit** [comment]
I suggest the following UI (if everything else would be too invasive or confusing):

* If a subflow is shown in the editor then a dropdown will appear at the top of the context tab to select the current subflow instance to work with.
* This dropdown is by default initialized with the most recently selected subflow node of that type, so if I double-click a subflow node and then click "edit subflow template" I would immediately be able to view the context data of that subflow instance I am most likely interested in without any further action. Of course I can select another instance at any time.
*

**Comment by vdwpsmt** [comment]
Hi, 
hitting on a problem exactly as described here.  I have xx instances of a subflow, each with their own internal context data.  Now I noticed I can not access this data from the side bar.  Is this still an open issue?  If so, any suggestions for a workaround?  I use context data all around, but without access I have no idea how to do debugging.

**Comment by vdwpsmt** [comment]
Hi, bumping this issue again.  this is quite an annoying problem when trying to quickly debug a problem in a subflow instance.  Any chance this issue will be covered on a short term?  If not, any suggestion for a workaround?

**Comment by baxerus** [comment]
Is there any progress?

Would really be nice to at least have the subflows flow context shown as the nodes context as [fvbommel](https://github.com/fvbommel) proposed, when an instance of the subflow is selected in the node editor.
The change to the UI should be minimal (if at all), but the benefit for debugging is big (of course this can than be refined to be nicer, like showing "Subflow" instead of "Node", but this is all not necessary for a first, useful iteration!).

**Comment by prof7bit** [comment]
> If so, any suggestions for a workaround? I use context data all around, but without access I have no idea how to do debugging.

One way to debug in some limited way is to use the status output: In the Subflow template itself I can wire the status output to something that will output debug information, for example I can either

* use a status node within the subflow and wire it to the status output to be able to mirror the status line of any node within my subflow to the outside without having to write code
* use a function node inside the subflow and within that produce a short line of debug text (maybe a short summary of parts of the subflow context) and write that to `node.status` and select that node in above mentioned status node
* or use any other means to rewrite the payload of a msg to contain the needed debug information in one short line of text and route this to the status output of the subflow.

That way you can have a line of text and a colored icon below each instance of the subflow that is updated in real time and reflects the current state of whatever you are currently trying to debug within that subflow.

This is currently the only way I have found to be able to look inside a running subflow instance to watch parts of its state change in realtime. Unfortunately you have to know exactly what you are interested in and set this up BEFORE the problem occurs that you want to debug.

**Comment by baxerus** [comment]
The idea of this feature request is to have insight of the state (actually the context) of an instance of a subflow, similar to how you can check the context of a (function) node.

My "workaround" for this is now to
1. only use flow context in the function nodes inside the subflow
1. Having a function node (I named it debug) in the signal flow directly after the input node of the subflow with the following code:
```
if (msg.debug_subflow !== undefined) {

    try {

        node.warn(flow.get(msg.debug_subflow))

    } catch (error) {
        
        const flow_keys = flow.keys()
        const debug_object = {}

        for (let key of flow_keys) {
            debug_object[key] = flow.get(key)
        }

        node.warn(debug_object)

    }

} else {

    return msg

}
```
(modify ⬆️ as you like)

1. Inject a message with a `debug_subflow` property to get a *warn message* (`node.debug` would probably be the right thing to do, but I want to see it on the side panel) in the log with:
    * **all** flow keys from the instances context when `debug_subflow` is set to a key that **does not exist** (e.g. use `true` or simply keep the default timestamp of the inject node)
    * only **one** key value form the instances context when `debug_subflow` is set to a key that **does exist** (can be helpful if you have lots of stuff in the context but are only interested in one in the moment)

Of course you can go all fancy with `change` and (eventually needed) `delay` nodes and a branched signal flow to get automatic log messages after the instance processes a message. But that totally depends on your needs and what fits best for you... 😉
